DESIGN.md

```diff
@@ -0,0 +1,108 @@
+
+### The WiX Compiler
+Vanagon uses three parts of the WiX compiler to construct an MSI from both the available WiX files and the components inside. As mentioned before, Vanagon stages all components in a staging directory. From this staging directory Vanagon executes the first WiX function: heat.exe. Heat will iterate over the entire staging directory and generate the necessary WiX file containing each file and directory under the staging directory.
+
+After heat Vanagon executes the compiler (candle.exe) then the linker (light.exe) on all WiX files staged to create the MSI. These two stages of WiX will happen on all files. Both those created dynamically by heat.exe / vanagon, and those static ones created by the user.
+
+### Windows Paths
+Before diving into the specifics of the WiX compiler it’s important to understand the windows pathing paradigm that WiX uses for the installation location of MSIs.
+
+Inside Vanagon::Platform::Windows.generate_msi_package, vanagon uses specific names for the stagedir. Specifically the SourceDir directory is important, because SourceDir is an MSI naming convention. The name refers to the “root” of the source files to be included in install. The idea behind the “root” is it provides the user creating an MSI a link to where the software installs to. On install, the MSI will recreate the entire structure (starting from SourceDir) in the install location, which will default to the largest drive, i.e. something like C:\. A good explanation for this is [here](http://stackoverflow.com/questions/1641094/in-wix-files-what-does-name-sourcedir-refer-to)
+
+Other than SourceDir the other related naming convention is ProgramFilesFolder / ProgramFiles64Folder. These two names are [MSI defaults](https://msdn.microsoft.com/en-us/library/windows/desktop/aa370905(v=vs.85).aspx) that allow us to utilize the localization engine for MSIs. On install these folders will resolve to the actual program files standard for windows, i.e. Program Files or Program Files (x86).
+
+### Heat.exe
+The heat process generates appropriate WiX elements corresponding to each file and directory under the stagedir. Vanagon uses heat’s output alongside the other static WiX files to include all files under the stagedir in the final MSI.
+Heat uses the SourceDir naming convention via input parameters to the command line. The process by which vanagon uses heat differs from the normal usage. This is because we cannot run heat anything in the CommonAppData folder. It’s likely the files/directories that go there need to be permanent and heat cannot handle permanent files/directories. In order to work around this we do not run heat.exe from SourceDir, as we only want to collect those files NOT under CommonAppData. Instead we run heat.exe from the path:
```

Comment:
Maybe re-word to say that some of the directories and files are added separately to the heat/harvest operation, so we only do the harvest under the program directory.
This also allows us to do additional special handling for the commonAppData (e.g. make sure configuration files are not over-written).
The harvest operation is like a bulk collection of multiple files which are handled in the default way (if that makes sense) - all of these files are treated as simple single file components with no additional handling.
Also mention that service files are filtered out from the harvest operation as they need additional handling as service component files.